airbyte-cdk/python/airbyte_cdk/destinations/vector_db_based/document_processor.py

```diff
@@ -0,0 +1,118 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from dataclasses import dataclass
+from typing import Any, Dict, List, Mapping, Optional, Tuple, Union
+
+import dpath.util
+from airbyte_cdk.destinations.vector_db_based.config import ProcessingConfigModel
+from airbyte_cdk.models import AirbyteRecordMessage, AirbyteStream, ConfiguredAirbyteCatalog, ConfiguredAirbyteStream, DestinationSyncMode
+from langchain.document_loaders.base import Document
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+from langchain.utils import stringify_dict
+
+METADATA_STREAM_FIELD = "_airbyte_stream"
+METADATA_RECORD_ID_FIELD = "_record_id"
+
+
+@dataclass
+class Chunk:
+    page_content: str
+    metadata: Dict[str, Any]
+    stream: str
+    namespace: Optional[str] = None
+
+
+class DocumentProcessor:
+    """
+    DocumentProcessor is a helper class that generates documents from Airbyte records.
+    It is used to generate documents from records before writing them to the destination:
+    * The text fields are extracted from the record and concatenated to a single string.
+    * The metadata fields are extracted from the record and added to the document metadata.
+    * The document is split into chunks of a given size using a langchain text splitter.
+
+    The Writer class uses the DocumentProcessor class to internally generate documents from records - in most cases you don't need to use it directly,
+    except if you want to implement a custom writer.
+
+    The config parameters specified by the ProcessingConfigModel has to be made part of the connector spec to allow the user to configure the document processor.
+    """
+
+    streams: Mapping[str, ConfiguredAirbyteStream]
+
+    def __init__(self, config: ProcessingConfigModel, catalog: ConfiguredAirbyteCatalog):
```

Comment:
is the `max_metadata_size` used in t[he langchain destination](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-langchain/destination_langchain/document_processor.py#L24) not needed?

Same question for the [metadata truncation](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-langchain/destination_langchain/document_processor.py#L81). Is that only needed for langchain?

Reply:
Very good point, thanks for sanity-checking this :) This is what I meant by
> Remove max_metadata logic and normalization for complex objects - as this is db-specific, the indexer should handle it appropriately

in the PR description - it was a mistake to put this into the document processor as it's an indexer concern (I actually had to extend it a little bit already on the indexer level for Chroma because it doesn't support boolean values: https://github.com/airbytehq/airbyte/blob/dd170e2e9dd0aa09110ae115657da1c83761f87b/airbyte-integrations/connectors/destination-langchain/destination_langchain/indexer.py#L195 ). I don't see any patterns yet in terms of normalizing the metadata that could be pulled to the general level - if something emerges, we can refactor it down similar to the custom low-code components.